### PR TITLE
smoke-b2b-Company-Account-in-Zed

### PR DESCRIPTION
**Pull request by durgesh-verma-spryker**
## PR Description
https://spryker.atlassian.net/browse/CC-21651

Company Account in Zed
- Smoke-b2b-Company-Account-in-Zed
- Create a new company in Zed, approve and activate
- Create a new BU to the new company
- Create a role for the new company
- Add new user to company, assign admin role
- Assign customer to the company (Customers → Customers→ Action "Attach to company")
- Assign the customer to another BU of the company (Company User → Action "Attach to BU")
- Add Company Business Unit Address

### CODE REVIEW

File: resources/steps/company_steps.robot [hunk 1]
```diff
@@ -1,110 +1,149 @@
...
-
-
+Resource   ../common/common.robot
+Resource   ../common/common_yves.robot
+Resource   ../common/common_zed.robot
+Resource   ../pages/zed/zed_attach_to_business_unit_page.robot
+Resource   ../pages/yves/yves_customer_account_page.robot
+Resource   ../pages/zed/zed_delete_company_user_page.robot
+ 
+*** Variables ***
```
Comment (AntonShopin) [inline resources/steps/company_steps.robot:9]: please don't describe variables in steps. Variables should be in pages. 
Check `resources/pages/zed/zed_create_company_business_unit_page.robot` `resources/pages/zed/zed_create_company_user_page.robot` and other pages ...

File: resources/steps/company_steps.robot [hunk 2]
```diff
@@ -1,110 +1,149 @@
...
+${zed_role_company_dropdown_locator}    id=company_role_create_form_fkCompany
+${zed_create_company_user_email_field}    id=company-user_customer_email
+${zed_create__company_user_salutation_dropdown}    id=company-user_customer_salutation
+${zed_create_company_user_first_name_field}    id=company-user_customer_first_name
+${zed_create_company_user_last_name_field}    id=company-user_customer_last_name
+${zed_create_company_user_gender_dropdow}    id=company-user_customer_gender
+${zed_create_company_user_dob_picker}    id=company-user_customer_date_of_birth
+${zed_create_company_user_phone_field}    id=company-user_customer_phone
+${zed_create_company_company_name_dropdown}    xpath=(//b[@role="presentation"])[1]
+${zed_create_company_business_unit_dropdown}    xpath=(//b[@role="presentation"])[2]
```
Comment (AntonShopin) [inline resources/steps/company_steps.robot:25]: looks like this locator needs to be updated. I can't get from where it's it and why you use role=presentation for 2 different locators, when one should be for company selector and the second for BU selector

File: resources/steps/company_steps.robot [hunk 3]
```diff
@@ -1,110 +1,149 @@
...
+${zed_save_button}    xpath=//input[@value="Save"]
```
Comment (AntonShopin) [inline resources/steps/company_steps.robot:26]: What is wrong with the already existing keyword: `Zed: submit the form`? Why do we need this new locator?

File: resources/steps/company_steps.robot [hunk 4]
```diff
@@ -1,110 +1,149 @@
...
+   [Documentation]     Creates new company BU with provided BU Name and for provided company.
+   [Arguments]    ${bu_name_to_set}    ${company_business_unit}
+   Zed: go to second navigation item level:    Customers    Company Units
+   Zed: click button in Header:    Create Company Business Unit
+   select from list by label    ${zed_bu_company_dropdown_locator}    ${company_business_unit}   
+   Type Text    ${zed_bu_name_field}    ${bu_name_to_set}
+   Type Text    ${zed_bu_iban_field}    testiban+${random}
+   Type Text    ${zed_bu_bic_field}    testbic+${random}
+   Zed: submit the form
+   # wait until element is visible    ${zed_success_flash_message}
```
Comment (AntonShopin) [inline resources/steps/company_steps.robot:39]: if this wait is not needed - remove, not comment

File: resources/steps/company_steps.robot [hunk 5]
```diff
@@ -1,110 +1,149 @@
...
+   IF    '/customer' not in '${currentURL}'    Zed: go to second navigation item level:    Customers    Company Users
+   Zed: perform search by:    ${companyUserName}
+   ${customerExists}=    Run Keyword And Return Status    Table should contain    ${zed_table_locator}    ${companyBusinessUnit}
+   IF    '${customerExists}'=='True'
+       Run keywords
+           Zed: click Action Button(without search) in a table for row that contains:    ${companyBusinessUnit}    Delete
+           Click    ${zed_confirm_delete_company_user_button}
+   END
+ 
+Zed: Business Unit Address
```
Comment (AntonShopin) [inline resources/steps/company_steps.robot:138]: Your keyword sounds like a statement, rename it to have a command. 

Keyword should support arguments

File: resources/steps/company_steps.robot [hunk 6]
```diff
@@ -1,110 +1,149 @@
...
+   Select From List By Text    id=company_unit_address_fkCompany    meta
+   Select From List By Label    id=company_unit_address_fkCountry    France
+   Type Text    id=company_unit_address_city    singapore
+   Type Text    id=company_unit_address_zipCode    ${random}
+   Type Text    id=company_unit_address_address1    address1${random}
```
Comment (AntonShopin) [inline resources/steps/company_steps.robot:143]: Why do you hardcode locators? They should be moved to pages

File: resources/steps/company_steps.robot [hunk 7]
```diff
@@ -1,110 +1,149 @@
...
+   Zed: submit the form
+ 
+Zed: Delete Company Business Unit:
```
Comment (AntonShopin) [inline resources/steps/company_steps.robot:146]: Please rename to: `Zed: delete company business unit:`

File: tests/smoke/smoke_b2b.robot [hunk 8]
```diff
@@ -26,844 +26,863 @@ Resource    ../../resources/steps/zed_availability_steps.robot
...
+    Zed: create new Company Business Unit with provided name and company:    Meta12BU    meta    
+    Zed: create new Company Role with provided permissions:    meta    Meta12role    true    
+    Zed: Create new Company User with provided email/company/business unit and role(s):    abc${random}@mail.com    meta    Headquarters    ADMINISTRATOR
+    Zed: go to second navigation item level:    Customers    Customers
+    Zed: click Action Button in a table for row that contains:    abc${random}@mail.com  Attach to company
+    Zed: go to second navigation item level:    Customers    Company Users
+    Zed: click Action Button in a table for row that contains:    Test  Attach to BU
+    Zed: go to second navigation item level:    Customers    Company Unit Addresses
+    Zed: click button in Header:    Create Company Unit Address
+    Zed: Business Unit Address
```
Comment (AntonShopin) [inline tests/smoke/smoke_b2b.robot:886]: What about this assertion? 
`Login in Yves as the new customer of the company → Go to company account details, check all details and users in Yves.`
It's the final step from https://spryker.atlassian.net/wiki/spaces/PS/pages/846791112/B2B+Demoshop+Smoke+Test+Checklist#B2BDemoshopSmokeTestChecklist-CompanyAccountinZed
If you don't know how to get the email - try a workaround with DataBase

**Review by AntonShopin — CHANGES_REQUESTED** [review]
Be aware that you have commented all tests in the smoke file!

File: resources/steps/company_steps.robot (hunk 1 above)
Comment (durgesh-verma-spryker) [reply resources/steps/company_steps.robot]: Fixed

File: resources/steps/company_steps.robot (hunk 2 above)
Comment (durgesh-verma-spryker) [reply resources/steps/company_steps.robot]: Fixed

File: resources/steps/company_steps.robot (hunk 5 above)
Comment (durgesh-verma-spryker) [reply resources/steps/company_steps.robot]: Fixed

File: resources/steps/company_steps.robot (hunk 7 above)
Comment (durgesh-verma-spryker) [reply resources/steps/company_steps.robot]: Fixed

File: tests/smoke/smoke_b2b.robot (hunk 8 above)
Comment (durgesh-verma-spryker) [reply tests/smoke/smoke_b2b.robot]: Fixed

File: resources/common/common_yves.robot [hunk 9]
```diff
@@ -298,3 +307,79 @@ Yves: page should contain script with attribute:
...
+    ${hasValue}    Run Keyword and return status     Should not be empty    ${headers}
+    ${response}=    IF    ${hasValue}   run keyword    PATCH   ${current_url}${path}    json=${data}    headers=${headers}    timeout=${timeout}    allow_redirects=${allow_redirects}    auth=${auth}    expected_status=${expected_status}
+    ...    ELSE    PATCH    ${current_url}${path}    json=${data}    timeout=${timeout}    allow_redirects=${allow_redirects}    auth=${auth}    expected_status=${expected_status}
+    ${response_body}=    IF    ${response.status_code} != 204    Set Variable    ${response.json()}
+    ${response_headers}=    Set Variable    ${response.headers}
+    Set Test Variable    ${response_headers}    ${response_headers}
+    Set Test Variable    ${response_body}    ${response_body}
+    Set Test Variable    ${response}    ${response}
+    Set Test Variable    ${expected_self_link}    ${current_url}${path}
+    [Return]    ${response_body}
```
Comment (AntonShopin) [inline resources/common/common_yves.robot:385]: Why API keywords are in UI keywords? If you need yo use API requests -> just add common_api in files (where it's needed) as a Resource

File: resources/common/common_yves.robot [hunk 10]
```diff
@@ -298,3 +307,79 @@ Yves: page should contain script with attribute:
...
+    ${hasValue}    Run Keyword and return status     Should not be empty    ${headers}
+    ${response}=    IF    ${hasValue}   run keyword    POST    ${current_url}${path}    json=${data}    headers=${headers}    timeout=${timeout}    allow_redirects=${allow_redirects}    auth=${auth}    expected_status=${expected_status}
+    ...    ELSE    POST    ${current_url}${path}    json=${data}    timeout=${timeout}    allow_redirects=${allow_redirects}    auth=${auth}    expected_status=ANY
+    ${response_body}=    IF    ${response.status_code} != 204    Set Variable    ${response.json()}
+    ${response_headers}=    Set Variable    ${response.headers}
+    Set Test Variable    ${response_headers}    ${response_headers}
+    Set Test Variable    ${response_body}    ${response_body}
+    Set Test Variable    ${response}    ${response}
+    Set Test Variable    ${expected_self_link}    ${current_url}${path}
+    [Return]    ${response_body}
```
Comment (AntonShopin) [inline resources/common/common_yves.robot:356]: Why API keywords are in UI keywords? If you need yo use API requests -> just add common_api in files (where it's needed) as a Resource

File: resources/steps/company_steps.robot [hunk 11]
```diff
@@ -1,110 +1,147 @@
...
+   Zed: perform search by:    ${companyUserName}
+   ${customerExists}=    Run Keyword And Return Status    Table should contain    ${zed_table_locator}    ${companyBusinessUnit}
+   IF    '${customerExists}'=='True'
+       Run keywords
+           Zed: click Action Button(without search) in a table for row that contains:    ${companyBusinessUnit}    Delete
+           Click    ${zed_confirm_delete_company_user_button}
+   END
+ 
+Zed: add business unit address:
+   [Arguments]    ${company}    ${country}    ${city}    ${ziprode}    ${street}
```
Comment (AntonShopin) [inline resources/steps/company_steps.robot:123]: typo in `ziprode`

File: resources/steps/company_steps.robot [hunk 12]
```diff
@@ -1,110 +1,147 @@
...
+   Select From List By Text    ${zed_comapny_unit_address_company_dropdown_locator}    meta
+   Select From List By Label    ${zed_comapny_unit_address_country_dropdown_locator}    France
+   Type Text    ${zed_company_unit_address_city_input_field_locator}    singapore
+   Type Text    ${zed_company_unit_address_zipcode_input_field_locator}    ${random}
+   Type Text    ${zed_company_unit_address_street_input_field_locator}    address1${random}
+   Zed: submit the form
```
Comment (AntonShopin) [inline resources/steps/company_steps.robot:129]: you expect arguments but they are not used inside the keyword. You hardcoded values

File: resources/steps/company_steps.robot [hunk 13]
```diff
@@ -1,110 +1,147 @@
...
+ 
+Zed: delete company business unit:
+   [Arguments]    ${name}
+   Zed: go to second navigation item level:    Customers    Company Units
+   Zed: click Action Button in a table for row that contains:    ${name}   Delete
+
+Yves: validate and set password for newly created company user:
+    [Arguments]    ${email}
```
Comment (AntonShopin) [inline resources/steps/company_steps.robot:137]: please add a second argument, for the password, it should not be hardcoded. You can specify the default value, but please don't hardcode. Example:
`Arguments]    ${email}    ${password}=${yves_user_password}` and check it PATCH request change ${yves_user_password} on ${password}

File: resources/steps/company_steps.robot [hunk 14]
```diff
@@ -1,110 +1,147 @@
...
+   Save the result of a SELECT DB query to a variable:    select registration_key from spy_customer where email = '${email}'    confirmation_key
+   Save the result of a SELECT DB query to a variable:    select customer_reference from spy_customer where email = '${email}'    customer_id
+   I send a POST request:     /customer-confirmation   {"data":{"type":"customer-confirmation","attributes":{"registrationKey":"${confirmation_key}"}}}
+   Save the result of a SELECT DB query to a variable:    select restore_password_key from spy_customer where email = '${email}'    restore_key
+   I send a PATCH request:    /customer-restore-password/${customer_id}   {"data":{"type":"customer-restore-password","id":"${customer_id}","attributes":{"restorePasswordKey":"${restore_key}","password":"${yves_user_password}","confirmPassword":"${yves_user_password}"}}}
+
+Yves: check details of users:
```
Comment (AntonShopin) [inline resources/steps/company_steps.robot:144]: please rename it and refactor. 
It should be self-explaining. 
Also, what does it check? Looks like the table with Users contains some specific name1 and name2. 
What are `name1 and name2`?

What if the table will contain 100 users, where name1 and name2 can be duplicated for different users. 

Also, why so bad locators? `//strong[normalize-space()='${name1}'] ` Just for example: `//div[@data-qa='component user-table']//table//td[@data-content='Name']/*[contains(text(),'${name1}')]`

File: tests/smoke/smoke_b2b.robot [hunk 15]
```diff
@@ -866,4 +866,28 @@ Refunds
...
     Zed: grand total for the order equals:    ${lastPlacedOrder}    €0.00
     [Teardown]    Run keywords    Zed: login on Zed with provided credentials:    ${zed_admin_email}
-    ...    AND    Zed: activate following discounts from Overview page:    20% off storage    10% off minimum order
\ No newline at end of file
+    ...    AND    Zed: activate following discounts from Overview page:    20% off storage    10% off minimum order
+
+Company_Account_in_Zed
+    [Documentation]    creating new company, BU, company role, company user and validate user details
+    Zed: login on Zed with provided credentials:    ${zed_admin_email}     
+    Zed: create new Company with provided name:    ${random}
```
Comment (AntonShopin) [inline tests/smoke/smoke_b2b.robot:874]: `${random}` is a company name, are you sure?

File: tests/smoke/smoke_b2b.robot [hunk 16]
```diff
@@ -866,4 +866,28 @@ Refunds
...
+    Zed: click Action Button in a table for row that contains:    ${random}   Activate
+    Reload
+    Zed: click Action Button in a table for row that contains:    ${random}   Approve
+    Zed: create new Company Business Unit with provided name and company:    Meta12BU    meta
```
Comment (AntonShopin) [inline tests/smoke/smoke_b2b.robot:878]: always add random to have unique companies, business units and so on. Maybe `Meta12BU${random}`?

Also, how does it work? First, You create a company with the name `${random}`, then you create a BU for a company with the name `meta`? How is it possible 

**Please answer in this comment on the following question: Do you run your tests before creating a PR?**

File: tests/smoke/smoke_b2b.robot [hunk 17]
```diff
@@ -866,4 +866,28 @@ Refunds
...
+
+Company_Account_in_Zed
+    [Documentation]    creating new company, BU, company role, company user and validate user details
+    Zed: login on Zed with provided credentials:    ${zed_admin_email}     
+    Zed: create new Company with provided name:    ${random}
+    Zed: click Action Button in a table for row that contains:    ${random}   Activate
+    Reload
+    Zed: click Action Button in a table for row that contains:    ${random}   Approve
+    Zed: create new Company Business Unit with provided name and company:    Meta12BU    meta    
+    Zed: create new Company Role with provided permissions:    meta    Meta12role    true
```
Comment (AntonShopin) [inline tests/smoke/smoke_b2b.robot:879]: Where are the permissions? You have no specified them

File: tests/smoke/smoke_b2b.robot [hunk 18]
```diff
@@ -866,4 +866,28 @@ Refunds
...
+Company_Account_in_Zed
+    [Documentation]    creating new company, BU, company role, company user and validate user details
+    Zed: login on Zed with provided credentials:    ${zed_admin_email}     
+    Zed: create new Company with provided name:    ${random}
+    Zed: click Action Button in a table for row that contains:    ${random}   Activate
+    Reload
+    Zed: click Action Button in a table for row that contains:    ${random}   Approve
+    Zed: create new Company Business Unit with provided name and company:    Meta12BU    meta    
+    Zed: create new Company Role with provided permissions:    meta    Meta12role    true    
+    Zed: Create new Company User with provided email/company/business unit and role(s):    abc${random}@mail.com    spryker    spryker    ADMINISTRATOR
```
Comment (AntonShopin) [inline tests/smoke/smoke_b2b.robot:880]: what? Previously, you have created a new Company, then you have create a new BU, and now you DON'T use them. Then why you have created a new company and BU and role of you don't use them?

File: tests/smoke/smoke_b2b.robot [hunk 19]
```diff
@@ -866,4 +866,28 @@ Refunds
...
+    Zed: go to second navigation item level:    Customers    Company Users
+    Zed: click Action Button in a table for row that contains:    Test  Attach to BU
+    Zed: go to second navigation item level:    Customers    Company Unit Addresses
+    Zed: click button in Header:    Create Company Unit Address
+    Zed: add business unit address:    meta    France    singapore    ${random}    address1${random}
+    Yves: login on Yves with provided credentials:    abc${random}@mail.com
+    Yves: go to company menu item:    Users
+    Yves: check details of users:    Robot First+${random}    Spryker Systems GmbH    
+    [Teardown]    Run Keywords    Zed: delete company user xxx withing xxx company business unit:    ${random}    spryker
+    ...    AND    Zed: delete company business unit:    ${random}
```
Comment (AntonShopin) [inline tests/smoke/smoke_b2b.robot:893]: I will not review this part until the previous comments are not fixed. No sense. 
Just a note, this test was 90% already developed and exists for > a year. Only part with setting a password for a newly created customer was missing. Read it, find it in the code and compare with yours

```UJ:Managing_B2B_Copmany_Account_in_Zed_Part_1
    [Documentation]    This user journey is to check possibility to Create Company, Business unit and initial Admin user with Permissions for Managing User in Zed
    Zed: Login on Zed with Provided Credentials:    ${admin_email}  ${default_password}
    Zed: create new Company with provided name:    Robot+${random}
    Zed: click Action Button in a table for row that contains:    Robot+${random}    Approve
    Zed: click Action Button in a table for row that contains:    Robot+${random}    Activate
    Zed: create new Company Role with provided permissions:    Robot+${random}    Test Robot Role+${random}    false    Add company users    Invite users    Enable / Disable company users    See Company Menu    Add item to cart    Change item in cart    Remove item from cart    Place Order    Alter Cart Up To Amount
    Zed: create new Company Role with provided permissions:    Trial    Existing Role+${random}    false    Add company users    Invite users    Enable / Disable company users    See Company Menu    Add item to cart    Change item in cart    Remove item from cart    Place Order    Alter Cart Up To Amount
    Zed: create new Company Business Unit with provided name and company:    Test Robot BU+${random}    Robot+${random}
    Zed: Create new Company User with provided email/company/business unit and role(s):    ${test_customer_email}    Robot+${random}    Test Robot BU+${random}    (id: ${newly_created_business_unit_id})    Test Robot Role+${random}
    Yves: Login on Yves with Provided Credentials:     ${test_customer_email}
    Yves: company menu 'should' be available for logged in user   
    Zed: click button in Header:  button_name```

File: resources/common/common_yves.robot (hunk 9 above)
Comment (durgesh-verma-spryker) [reply resources/common/common_yves.robot:385]: Done, keeping as it is , as updating it leads to conflict in setup

File: resources/common/common_yves.robot (hunk 10 above)
Comment (durgesh-verma-spryker) [reply resources/common/common_yves.robot:356]: Done, keeping as it is , as updating it leads to conflict in setup

File: tests/smoke/smoke_b2b.robot (hunk 18 above)
Comment (durgesh-verma-spryker) [reply tests/smoke/smoke_b2b.robot]: Fixed

File: tests/smoke/smoke_b2b.robot (hunk 17 above)
Comment (durgesh-verma-spryker) [reply tests/smoke/smoke_b2b.robot]: Fixed

File: tests/smoke/smoke_b2b.robot (hunk 15 above)
Comment (durgesh-verma-spryker) [reply tests/smoke/smoke_b2b.robot]: Fixed, as per your suggestion

File: resources/steps/company_steps.robot (hunk 14 above)
Comment (durgesh-verma-spryker) [reply resources/steps/company_steps.robot]: Fixed

File: resources/steps/company_steps.robot (hunk 13 above)
Comment (durgesh-verma-spryker) [reply resources/steps/company_steps.robot]: fixed

File: resources/steps/company_steps.robot (hunk 12 above)
Comment (durgesh-verma-spryker) [reply resources/steps/company_steps.robot]: Fixed

File: resources/steps/company_steps.robot (hunk 11 above)
Comment (durgesh-verma-spryker) [reply resources/steps/company_steps.robot]: Fixed

File: resources/common/common_yves.robot (hunk 10 above)
Comment (durgesh-verma-spryker) [reply resources/common/common_yves.robot:356]: Done, keeping as it is , as updating it leads to conflict in setup

File: resources/common/common_yves.robot (hunk 9 above)
Comment (durgesh-verma-spryker) [reply resources/common/common_yves.robot:385]: Done, keeping as it is , as updating it leads to conflict in setup

File: tests/smoke/smoke_b2b.robot (hunk 19 above)
Comment (durgesh-verma-spryker) [reply tests/smoke/smoke_b2b.robot]: All the previous comments are fixed

File: resources/common/common_yves.robot (hunk 10 above)
Comment (AntonShopin) [reply resources/common/common_yves.robot:356]: What is the reason why you still keep it in common_yves?

File: resources/common/common_yves.robot (hunk 9 above)
Comment (AntonShopin) [reply resources/common/common_yves.robot:385]: What is the reason why you still keep it in common_yves?

File: tests/smoke/smoke_b2b.robot [hunk 20]
```diff
@@ -866,4 +866,27 @@ Refunds
...
+    Zed: click Action Button in a table for row that contains:    ${test_customer_email}  Attach to company
+    Zed: go to second navigation item level:    Customers    Company Users
+    Zed: click Action Button in a table for row that contains:    Robot+${random}  Attach to BU
+    Zed: go to second navigation item level:    Customers    Company Unit Addresses
+    Zed: click button in Header:    Create Company Unit Address
+    Zed: add business unit address:    Robot+${random}    France    singapore    ${random}    address1${random}
+    Yves: login on Yves with provided credentials:    ${test_customer_email}
+    Yves: go to company menu item:    Users
+    Yves: check the user info of company users:    Robot First+${random}    Test Robot BU+${random}
+    [Teardown]    Run Keywords    Zed: delete company user xxx withing xxx company business unit:    ${random}    spryker
```
[the remark this answers is not in the document]
Comment (durgesh-verma-spryker) [reply tests/smoke/smoke_b2b.robot]: Fixed, Mistakenly Forgot replace with '${test_customer_email}' ,'Test Robot BU+${random}'

File: resources/common/common_yves.robot (hunk 10 above)
Comment (durgesh-verma-spryker) [reply resources/common/common_yves.robot:356]: image

when I add a common_api file as a resource and remove these keywords from common_yves. I get conflict in suite setup and I am not able to execute tests. That's the reason I did not remove it.

**Pull request closed without merging**